support/langkit_support-generic_api-analysis.ads

```diff
@@ -0,0 +1,270 @@
+-- granted additional  permissions described in the  GCC  Runtime  Library  --
+-- Exception, version 3.1, as published by the Free Software Foundation.    --
+--                                                                          --
+-- You should have received a copy of the GNU General Public License and a  --
+-- copy of the GCC Runtime Library Exception along with this program;  see  --
+-- the files COPYING3 and COPYING.RUNTIME respectively.  If not, see        --
+-- <http://www.gnu.org/licenses/>.                                          --
+------------------------------------------------------------------------------
+
+--  This package provides a generic API so that programs can work with the
+--  $.Analysis packages of all Langkit-generated libraries.
+--
+--  Note that it is experimental at this stage, and thus not officially
+--  supported.
+
+private with Ada.Finalization;
+
+with Langkit_Support.File_Readers; use Langkit_Support.File_Readers;
+with Langkit_Support.Internal;     use Langkit_Support.Internal;
+with Langkit_Support.Text;         use Langkit_Support.Text;
+
+package Langkit_Support.Generic_API.Analysis is
+
+   type Analysis_Context is tagged private;
+   type Analysis_Unit is new Langkit_Support.Text.Text_Buffer_Ifc with private;
+   type Analysis_Node is tagged private;
+   --  Reference types to actual context/unit/node objects. Only
+   --  Analysis_Context and Analysis_Unit are strong references (designated
+   --  object lives as long as the reference exists): Analysis_Node is a weak
+   --  reference (designated object may be destroyed even if there are still
+   --  references to it, but usage of such a stale reference is properly
+   --  rejected).
+
+   No_Analysis_Context : constant Analysis_Context;
+   --  Special value to mean the absence of analysis context
+
+   No_Analysis_Unit : constant Analysis_Unit;
+   --  Special value to mean the absence of analysis unit
+
+   No_Analysis_Node : constant Analysis_Node;
+   --  Special value to mean the absence of analysis node
+
+   --  TODO: add unit providers handling
+
+   ---------------------------------
+   -- Analysis context operations --
+   ---------------------------------
+
+   function Create_Context
+     (Language      : Language_Id;
+      Charset       : String := "";
+      File_Reader   : File_Reader_Reference := No_File_Reader_Reference;
+      With_Trivia   : Boolean := True;
+      Tab_Stop      : Natural := 0)
+      return Analysis_Context;
+   --  Create a new analysis context.
+   --
+   --  ``Charset`` will be used as a default charset to decode input sources
+   --  in analysis units. Please see ``GNATCOLL.Iconv`` for several supported
+   --  charsets. Be careful: passing an unsupported charset is not guaranteed
+   --  to raise an error here. If no charset is provided, use the language
+   --  default.
+   --
+   --  When ``With_Trivia`` is true, the parsed analysis units will contain
```

Comment:
We're probably missing a definition of trivia in user doc. Can you put at least a TODO here ? I'll see if there are other language agnostic concepts that we don't explain

Reply:
Trivias are defined in LAL’s manual (https://docs.adacore.com/live/wave/libadalang/html/libadalang_ug/core_concepts.html#tokens-and-trivias) but also in Langkit_Support (https://github.com/AdaCore/langkit/blob/master/support/langkit_support-token_data_handlers.ads#L53). What would you suggest in addition?

Reply:
I'm not sure what I suggest TBH, but clearly the current situation is not great from a langkit perspective. We're lacking an explanation of some general langkit concepts, and if you try to use langkit libraries generically, you might not think of going into LAL's user guide, nor in token_data_handlers.ads line 53

Reply:
Ok, I’ll add a general TODO about this at the top of this unit